src/nnpack/interface.jl

```diff
@@ -0,0 +1,139 @@
+function check_support(x, k, pad, stride, dilation = 1)
+    fallback = false
+    dilation == 1 || dilation == (1, 1) || (fallback = true)
+    pad_, stride_ = expand(Val{length(k)}, pad), expand(Val{length(k)}, stride)
+    ((size(x, 1) - k[1] + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - k[2] + 2 * pad_[2]) % stride_[2] == 0) || (fallback = true)
+    return pad_, stride_, fallback
+end
+
+function softmax!(x::A) where A<:AbstractVecOrMat{Float64}
+    x = Float32.(x)
+    softmax!(x)
+end
+
+softmax!(x::A) where A<:AbstractVecOrMat{Float32} =
+    nnp_softmax_output(x, x, threadpool = shared_threadpool[])
+
+softmax!(y::A, x::A) where A<:AbstractVecOrMat{Float64} = softmax!(Float32.(y), Float32.(x))
+
+softmax!(y::A, x::A) where A<:AbstractVecOrMat{Float32} =
+    nnp_softmax_output(x, y, threadpool = shared_threadpool[])
+
+softmax(x::A) where A<:AbstractVecOrMat{Float64} = softmax(Float32.(x))
+
+softmax(x::A) where A<:AbstractVecOrMat{Float32} =
+    nnp_softmax_output(x, similar(x), threadpool = shared_threadpool[])
+
+maxpool(x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float64, 4} =
+    maxpool(Float32.(x), k, pad = pad, stride = stride)
+
+function maxpool(x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float32, 4}
+    pad_, stride_, fallback = check_support(x, k, pad, stride)
+    if fallback
+        maxpool_cpu!(similar(x, pdims(size(x), k, pad_, stride_)), x, k, pad = pad_, stride = stride_)
+    else
+        maxpool!(similar(x, pdims(size(x), k, pad_, stride_)), x, k, pad = pad_, stride = stride_)
+    end
+end
+
+maxpool!(y::A, x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float64, 4} =
+    maxpool!(Float32.(y), Float32.(x), k, pad = pad, stride = stride)
+
+maxpool!(y::A, x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float32, 4} =
```

Comment:
Is AbstractArray really right here? Can you actually pass anything other than `Array` to nnpack?